lib/subscribers/google-genai/base.js

```diff
@@ -0,0 +1,127 @@
+ */
+
+const {
+  LlmChatCompletionMessage,
+  LlmChatCompletionSummary,
+  LlmErrorMessage
+} = require('../../../lib/llm-events/google-genai')
+const { extractLlmContext } = require('../../util/llm-utils')
+const { AI } = require('../../../lib/metrics/names')
+const { GEMINI } = AI
+const { DESTINATIONS } = require('../../config/attribute-filter')
+
+const Subscriber = require('../base')
+
+class GoogleGenAISubscriber extends Subscriber {
+  constructor({ agent, logger, channelName, packageName = '@google/genai' }) {
+    super({ agent, logger, channelName, packageName })
+    this.events = ['asyncEnd']
+  }
+
+  get enabled() {
+    return super.enabled && this.config?.ai_monitoring?.enabled
+  }
+
+  /**
+   * Enqueues a LLM event to the custom event aggregator
+   *
+   * @param {object} params input params
+   * @param {string} params.type LLM event type
+   * @param {object} params.msg LLM event
+   */
+  recordEvent({ type, msg }) {
+    const llmContext = extractLlmContext(this.agent)
+
+    this.agent.customEventAggregator.add([
+      { type, timestamp: Date.now() },
+      Object.assign({}, msg, llmContext)
+    ])
+  }
+
+  /**
+   * Generates LlmChatCompletionSummary for a chat completion creation.
+   * Also iterates over both input messages and the first response message
+   * and creates LlmChatCompletionMessage.
+   *
+   * Also assigns relevant ids by response id for LlmFeedbackEvent creation
+   *
+   * @param {object} params input params
+   * @param {TraceSegment} params.segment active segment from chat completion
+   * @param {object} params.request chat completion params
+   * @param {object} [params.response] chat completion response
+   * @param {boolean} [params.err] err if it exists
+   * @param {Transaction} params.transaction active transaction
+   */
+  recordChatCompletionMessages({
+    segment,
+    request,
+    response,
+    err,
+    transaction
+  }) {
+    const agent = this.agent
+    if (!response) {
+      // If we get an error, it is possible that `response = null`.
+      // In that case, we define it to be an empty object.
+      response = {}
+    }
+
+    // Explicitly end segment to provide consistent duration
+    // for both LLM events and the segment
+    segment.end()
+    const completionSummary = new LlmChatCompletionSummary({
+      agent,
+      segment,
+      transaction,
+      request,
+      response,
+      withError: err != null
+    })
+
+    // Only take the first response message and append to input messages
+    // request.contents can be a string or an array of strings
+    // response.candidates is an array of candidates (choices); we only take the first one
+    const inputMessages = Array.isArray(request.contents) ? request.contents : [request.contents]
+    const responseMessage = response?.candidates?.[0]?.content
+    const messages = responseMessage !== undefined ? [...inputMessages, responseMessage] : inputMessages
+    messages.forEach((message, index) => {
```

Comment:
This was refactored to remove .`forEach`, [see](https://github.com/newrelic/node-newrelic/blob/f281f19fbc40f91ae80957ef8c6b98f1df23f606/lib/instrumentation/%40google/genai.js#L98)